zebra-network/src/peer_set/set.rs

```diff
@@ -187,13 +203,28 @@ where
             }
         }
 
-        match Pin::new(&mut self.guards).poll_next(cx) {
-            Poll::Pending => {}
-            Poll::Ready(Some(res)) => res??,
-            Poll::Ready(None) => Err("all background tasks have exited")?,
+        let exit_error = match Pin::new(&mut self.guards).poll_next(cx) {
+            Poll::Pending => return Ok(()),
+            Poll::Ready(Some(res)) => {
+                info!(
+                    background_tasks = %self.guards.len(),
+                    "a peer set background task exited, shutting down other peer set tasks"
+                );
+
+                // Flatten the join result and inner result,
+                // then turn Ok() task exits into errors.
+                res.map_err(Into::into)
+                    .and_then(convert::identity)
+                    .and(Err("a peer set background task exited".into()))
+            }
+            Poll::Ready(None) => Err("all peer set background tasks have exited".into()),
+        };
+
+        for guard in self.guards.iter() {
+            guard.abort();
```

Comment:
Question: Isn't this covered by `Drop`?

Reply:
> Question: Isn't this covered by `Drop`?

Nevermind, I looked into it and `Drop` just stops watching the task :+1:

Reply:
Actually, you're right, Drop should cancel these tasks as well.

Reply:
Fixed in PR #3078.